subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,146 +122,172 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
+	memcpy(slot->slot.metadata, metadata, metadata_len);
+	return 0;
+}
 
-	if (valid) {
-		prev = atomic_test_and_set_bit(valid_slots, idx);
-	} else {
-		prev = atomic_test_and_clear_bit(valid_slots, idx);
-	}
+int bt_mesh_dfu_slot_fwid_set(struct bt_mesh_dfu_slot *dfu_slot,
+			      const uint8_t *fwid, size_t fwid_len)
+{
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (valid == prev) {
-		return 0;
+	if (fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		return -EFBIG;
 	}
 
-	err = valid_slots_store();
-	if (err) {
-		LOG_WRN("Storage failed. err: %d", err);
-		atomic_set_bit_to(valid_slots, idx, prev);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	return err;
+	slot->slot.fwid_len = fwid_len;
+	memcpy(slot->slot.fwid, fwid, fwid_len);
+	return 0;
 }
 
-bool bt_mesh_dfu_slot_is_valid(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_commit(struct bt_mesh_dfu_slot *dfu_slot)
 {
-	uint16_t idx;
+	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_RESERVED ||
+	    slot->slot.fwid_len == 0 ||
+	    slot->slot.size == 0) {
+		return -EINVAL;
+	}
+
+	if (bt_mesh_dfu_slot_get(slot->slot.fwid, slot->slot.fwid_len, NULL)
+	    >= 0) {
+		return -EALREADY;
+	}
+
+	slot->idx = committed_slot_count;
+	slot->state = SLOT_STATE_COMMITTED;
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return false;
+	err = slot_store(slot);
+	if (err) {
+		LOG_WRN("Store failed (err: %d)", err);
+		slot->state = SLOT_STATE_RESERVED;
+		slot->idx = 0;
+		return err;
 	}
 
-	idx = slot_idx(slot);
-	return atomic_test_bit(valid_slots, idx);
+	committed_slot_count++;
+
+	LOG_DBG("Stored slot #%u: %s", slot - &slots[0],
+		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
+	return 0;
 }
 
-int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_release(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	struct slot *s = CONTAINER_OF(slot, struct slot, slot);
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	LOG_DBG("%u", s - &slots[0]);
-
-	slot_erase(s);
-	slot_invalidate(s);
-	sys_slist_find_and_remove(&list, &s->n);
+	slot->state = SLOT_STATE_FREE;
 
 	return 0;
 }
 
-int bt_mesh_dfu_slot_del_all(void)
+int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	struct slot *s;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_COMMITTED) {
+		return -EINVAL;
+	}
+
+	int idx = slot - &slots[0];
+
+	LOG_DBG("%u", idx);
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&list, s, n) {
-		slot_erase(s);
-		slot_invalidate(s);
+	slot_erase(slot);
+	slot->state = SLOT_STATE_FREE;
+
+	for (int i = 0; i < CONFIG_BT_MESH_DFU_SLOT_CNT; i++) {
+		if (slots[i].idx > slot->idx) {
+			slots[i].idx--;
+			slot_store(&slots[i]);
+		}
```

Comment:
Can we optimize this somehow? For example, use some 32bits global index for every new slot to identify slot related info (header, firmware id, metadata), but store index only in header. Then we don't need at least to store again all slot information every time we delete slot with lower index.

I could even suggest to simply use 32bits index which is never decremented. It may be a bit complicated to get correct indexes as we will need to skip deleted slots, but at least we won't need to do extra flash store here. One can argue that 32bits can wrap around eventually, but I guess flash will wear out earlier. Well, or add some fragmentation logic when index reaches UINT32_MAX.

Also just storing here is not enough because as the flash entry with
the highest slot index will still stay.

Reply:
> Also just storing here is not enough because as the flash entry with
the highest slot index will still stay.

No. The flash entries are indexed by the backing array index, not the firmware list index (i.e. `idx` field). It will just write new `idx` values to the same flash entries.

The point about storing is valid. I considered this to be passable, seeing as a) there are typically not that many slots and b) removing one of them at a time is not something that happens very often, so I decided on saving the complication of the code. Keep in mind, this will not store the contents of fwid and metadata again, so the amount written is not giant.

Reply:
I can't find in `slot_store` where it only stores header, but skips metadata and fwid. `idx` is also included in the path from what, so if idx here changes, all data is stored again. Or am I missing something>

Reply:
Oh no sorry you are right. It will store everything. My plan was to only store the header again, I seem to have forgotten to do that. Will that be enough in your opinion?

Reply:
And store the list nodes in flash? Or rebuild the list with the slow "look for next number" algorithm when loading?

Reply:
The counter will only be needed to restore the order of entries when loading. And it doesn't need to be stored separately. It can be restored after the loading from the last slot.

I don't think it will be slow in this case. I think it will be matter of tens of slots and this will happen only once at boot. And it solves all problems from the API side. We don't need to restore order and use extra stack space (https://github.com/zephyrproject-rtos/zephyr/pull/61389#discussion_r1300211715), removal and addition will be simpler, etc.

Reply:
What are we trying to optimize here, and is it worth doing it? The solution seems a bit more complex and time consuming than the problem it is trying to solve. There will never be more than handful of slots (unless if someone decides to use external 10MB/100MB/1GB flash (quite possible for gateways), and in such case, current solution should be fine in fact).

Reply:
I lean towards agreeing with Omkar. I see your worry about extensive flash usage as the main issue you are trying to solve, @PavelVPV ? If so, I think my preferred improvement here would be to optimize by only storing the header again for the slots that need to update their index after a slot removal. I agree with Omkar that bringing back the linked list seems a little overkill only to save flash usage we're not aware that is or will ever be a problem for anyone yet. Keep in mind that the bulk of the slot entry is the metadata and fwid data itself, so skipping a re-flash of these will save quite a bit of flash activity.

Reply:
Does this solves the issue: https://github.com/zephyrproject-rtos/zephyr/compare/main...PavelVPV:fix_dfu_slots_reordering?expand=1 ?

Does defragmentation part looks complex?

Reply:
After thinking about this a bit more, I would like to revise my stance on this. I think it may be worth prioritizing minimizing flash writes. For gateway applications, it will be possible that there will be several re-writes of slots due to complex multi-device network requiring FW updates, say once or twice in a month. DFU need not be just for firmware images, it could very well be used for calibrating light sources and updating LUTs and such things.

Reply:
So, everyone now agrees to reintroduce the linked list then? I see your point about defrag @PavelVPV , I hadn't thought it fully through.